inst/pages/98_exercises.qmd

```diff
@@ -1158,62 +1155,145 @@ See the OMA book for examples.
-pcoa_bray_clusters <- kmeans(reducedDim(pcoa_bray), centers = 3)$cluster
+#| label: heatmap-visualization-ex1
+#| eval: FALSE
+#| code-fold: true
+#| code-summary: "Show solution"
 
-# Plot the reduced dimensions of pcoa_bray with clusters as colors
-plotReducedDim(pcoa_bray, colour_by = pcoa_bray_clusters)
+# Load necessary libraries
+library(mia)
+library(SummarizedExperiment)
+library(vegan)
+library(bluster)
+library(scater)
+
+# Generate a sample SummarizedExperiment object
+data(GlobalPatterns)
+tse <- GlobalPatterns
 
-# Perform k-means clustering on the reduced dimensions of pca_aitchison
-pca_aitchison_clusters <- kmeans(reducedDim(pca_aitchison), centers = 3)$cluster
+# Perform MDS on the abundance data using scater::runMDS
+# Specify the correct assay to be used
+tse <- runMDS(tse, FUN = vegan::vegdist, method = "bray", assay.type = "counts")
 
-# Plot the reduced dimensions of pca_aitchison with clusters as colors
-plotReducedDim(pca_aitchison, colour_by = pca_aitchison_clusters)
+# Perform k-means clustering (k = 3) using the bluster package
+set.seed(123)
+clustering_result <- clusterRows(reducedDim(tse, "MDS"), BLUSPARAM=KmeansParam(centers=3))
+colData(tse)$cluster <- as.factor(clustering_result)
+
+# Visualize MDS with clusters
+plotReducedDim(tse, dimred = "MDS", colour_by = "cluster") +
+  labs(title = "MDS with k-means Clustering") +
+  theme_minimal() +
+  scale_color_manual(values = c("red", "blue", "green"))
 ```
-Where : 
-'reducedDim' is a function that extracts the reduced dimensionality representation of your data (e.g., from a PCA or 
-PCoA object).
-'plotReducedDim' is a function that can plot these reduced dimensions and has a 'colour_by argument' that accepts a vector of cluster assignments.
 
-#### 2. Heatmap with Annotations:
+### 2: Cluster Analysis with Heatmap Visualization
 
-* Annotate the heatmap with additional metadata (e.g., subject groups or sample types).
+- Hierarchical clustering of samples based on abundance data.
+- Visualizes the clustered data using a heatmap, with cluster information added to metadata.
 
 ```{r}
+#| label: heatmap-visualization-ex2
+#| eval: FALSE
+#| code-fold: true
+#| code-summary: "Show solution"
+
 # Load necessary libraries
+library(mia)
 library(SummarizedExperiment)
-library(pheatmap)
+library(vegan)
+library(bluster)
+library(scater)
+library(ComplexHeatmap)
+library(circlize)
+library(compositions)
+
+# Generate a sample SummarizedExperiment object 
+data(GlobalPatterns)
+tse <- GlobalPatterns
 
-# Example data creation
-# Assuming enterotype is a SummarizedExperiment object with sample metadata
-# and assay_data is a matrix containing assay data
+# Extract the abundance data matrix and perform CLR transformation
+clr_data <- transformAssay(se, method = "clr", assay = "counts")
+
+# Perform hierarchical clustering on the samples using Euclidean distance
+distance_matrix <- dist(clr_data, method = "euclidean")
```

Comment:
clr_data should be TreeSE object as it comes from transformAssay(). stats::dist requires matrix object which indicates that this line is not working.

Reply:
Please run the code and check that it is working

Reply:
It works when I build it on my local system.

Reply:
This clearly does not work. 

```
# Get example data
library(mia)
data(GlobalPatterns)
tse <- GlobalPatterns
se <- GlobalPatterns

# Try this, doesn't work:
clr_data <- transformAssay(se, method = "clr", assay = "counts")
#Error in transformAssay(se, method = "clr", assay = "counts") : 
# argument 3 matches multiple formal arguments

# Ok first issue is that there is no argument "assay" in transformAssay function.. so let's change it:
clr_data <- transformAssay(se, method = "clr", assay.type = "counts")
#Error: 'clr' cannot be used with non-positive data: use pseudocount > 0

# Ok pseudocount was also missing, let's add:
clr_data <- transformAssay(se, method = "clr", assay.type = "counts", pseudocount=TRUE)
# A pseudocount of 0.5 was applied.

# This also doesn't work for SE objects
distance_matrix <- dist(clr_data, method = "euclidean")
#Error in h(simpleError(msg, call)) : 
#  error in evaluating the argument 'x' in selecting a method for function 'as.matrix': no method for coercing #this S4 class to a vector

# It does work for matrix (assay) as Tuomas said:
distance_matrix <- dist(assay(clr_data, "clr"), method = "euclidean")
```

-> Run all code before pushing to PR